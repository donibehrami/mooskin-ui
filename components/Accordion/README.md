# MooSkin UI - Accordion Component

The MooSkin Accordion Component is like a tabbed window, similar to RadioTabs component, with as many radio tabs as you like.

___

### Usage

To start using the Accordion Component first you have to Import its two main components

```
Import {RadioTabs, Content} from 'mooskin';
```
or
```
import Accordion, {Content} from 'mooskin/lib/Accordion';
```

And then you can simply start using it by typing

```
<Accordion>
    <Content title="Title 1" active>
        <div style={{...style, backgroundColor: '#5CCDDF'}}>
            Content 1
        </div>
    </Content>
    <Content title="Title 2">
        <div style={{...style, backgroundColor: '#F48770'}}>
            Content 2
        </div>
    </Content>
    <Content title="Title 3">
        <div style={{...style, backgroundColor: '#F2C14A'}}>
            Content 3
        </div>
    </Content>
</Accordion>
```


Individual <Content> components has a required 'title' prop which will be displayed in the radio header, the rest are optional. You can put whatever inside <Content> tags, it will be displayed as content for that particular tab.

### Examples

Can be used without importing the <Content> component with shorthand dot(.) notation.

```
Import {Accordion} from 'mooskin';

<Accordion>
    <Accordion.Content title="Title 1" iconClass="fa fa-icon">
        <div style={{...style, backgroundColor: '#5CCDDF'}}>
            Content 1
        </div>
    </Accordion.Content>
    <Accordion.Content title="Title 2" materialIcon="home">
        <div style={{...style, backgroundColor: '#F48770'}}>
            Content 2
        </div>
    </Accordion.Content>
</Accordion>
```

```
Import {Content, Accordion} from 'mooskin';

<Accordion horizontal>
    <Content title="Title 1" iconClass="fa fa-icon">
        <div style={{...style, backgroundColor: '#5CCDDF'}}>
            Content 1
        </div>
    </Content>
    <Content title="Title 2" materialIcon="home">
        <div style={{...style, backgroundColor: '#F48770'}}>
            Content 2
        </div>
    </Content>
</Accordion>
```

## Supported attributes for ```<Accordion/>```

* `id` - id of the element
* `className` - css class
* `style` - extra styles for the main container


## Supported attributes for ```<Content/>```

* `title` - (required) sets the title for the radio tab
* `active` - sets which radio tab is active, the first is active by default
* `style` -  extra styles 

Allthough these attributes are supported, most of them are optional.

___

#### For more

[MooSkin-UI](https://github.com/moosend/mooskin-ui)