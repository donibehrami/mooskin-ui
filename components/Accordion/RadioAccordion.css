.container {
    display: grid;
    grid-column-template: 1fr;
    grid-gap: 5px;
}

.radio span {
    margin-left: 10px;
    cursor: pointer;
    vertical-align: middle;
    user-select: none;
}

.radio {
    border-radius: 8px;
    color: #5ccdde;
    background: transparent;
    transition: 0.3s;
    border: 1px solid #5ccdde;
}

.radio:hover {
    background-color: #95dee9;
    color: #fff;
    border-color: #95dee9;
}

.radio input[type='radio'] {
    vertical-align: middle;
    appearance: none;
    margin: 0 auto;
    cursor: pointer;
    width: 1.188rem;
    height: 1.188rem;
    padding: 0.3425rem;
    background: #fff;
    border-radius:50%;
    border: 1px solid #5ccdde;
}

.radio:hover input[type='radio']{
    border-color: #95dee9;
}

.radio input[type='radio']:checked::after {
    background:#5ccdde;
    content:'';
    display:block;
    width: 0.375rem;
    height: 0.375rem;
    border-radius:50%; 
}

.radio label {
    display: inline-block;    
    cursor: pointer;
    padding: 12px;
    height: 100%;
    display: block; 
    border-radius: 8px;
    font-size: 0.875rem;
    width: 100%;
    height: 100%;
}

.checkedRadio {
    background: #5ccdde;
    color: #fff;
}

.checkedRadio:hover {
    background: #5ccdde;
    border-color: #5ccdde;
}

.checkedRadio:hover input[type='radio']{
    border-color: #5ccdde;
}

.content {
    color: #7b7b7b;
    font-size: 0.875rem;
    animation-name: fadeInFromNone;
    animation-duration: 0.5s;
    padding: 10px;
}

.visible {
    display: block;
}

.invisible {
    display: none;
}

@keyframes fadeInFromNone {
    0% {
        display: none;
        opacity: 0;
    }

    1% {
        display: block;
        opacity: 0;
    }

    100% {
        display: block;
        opacity: 1;
    }
}