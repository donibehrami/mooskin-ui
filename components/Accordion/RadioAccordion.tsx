import * as React from 'react';

import styles from './RadioAccordion.css';

export interface IAccordionProps {

    /** override id */
    id?: string;

    /** container class */
    className?: string;

    /** override styles */
    style?: {[key: string]: string|number};

    /** children here can only be Tab elements */
    children?: Array<React.ReactElement<IContentProps>> | React.ReactElement<IContentProps>;
}

export interface IContentProps {

    /** container class */
    className?: string;

    /** title */
    title: string;

    /** if active */
    active?: boolean;

    /** override styles */
    style?: {[key: string]: string|number};

}

export interface IHeaderProps {
    title: string;
    active: boolean;
    name?: string;
    onClick: (e: React.MouseEvent<HTMLElement>) => void;
}

export interface IAccordionState {
    active?: number;
}

export default class RadioAccordion extends React.Component<IAccordionProps, IAccordionState>{

    public static defaultProps = {
        active: 0,
        className: '',
        style: {}
    };

    public static AccordionContent: React.StatelessComponent<IContentProps>;

    private name: string;

    constructor(props: IAccordionProps){
        super(props);

        this.name = this.generateName();

        this.state = {
            active: this.getActiveTab()
        };
    }

    public render() {

        const accordion = this.makeContent();

        return(
            <div className={`accordion-component ${styles.container}`}>
                {accordion}
            </div>
        );
    }

    public onClickHeader = (index: number) => {
        return (e: React.MouseEvent<HTMLElement>) => {
            this.setState({active: index});
        };
    }

    private makeContent(){

        const accordion: Array<React.ReactElement<IHeaderProps>> = [];

        React.Children.forEach(this.props.children, (child, index) => {
            if (React.isValidElement<IContentProps>(child)){

                const extraProps: Partial<IContentProps & {key: number}> = {
                    active: this.state.active === index,
                    key: index
                };

                accordion.push(
                    <Header
                        key={index}
                        title={child.props.title}
                        active={this.state.active === index}
                        onClick={this.onClickHeader(index)}
                        name={this.name}
                    >
                        {React.cloneElement(child, extraProps)}
                    </Header>
                );

            }else{
                throw new Error('<Accordion> element only accepts Content elements as children');
            }
        });

        return accordion;
    }

    private getActiveTab() {
        const childrenArray = React.Children.toArray(this.props.children);

        for (const [index, value] of childrenArray.entries()){
            if (React.isValidElement<IContentProps>(value) && value.props.active){
                return index;
            }
        }
    }

    private generateName = () => {
        return Date.now().toString();
    }
}

export const AccordionContent: React.StatelessComponent<IContentProps> = (props) => {

    const displayClass = !props.active ? styles.invisible : styles.visible;

    return (
        <div
            className={`accordion-content ${styles.content} ${displayClass}`}
            style={props.style}
        >
            {props.children}
        </div>
    );
};

export const Header: React.StatelessComponent<IHeaderProps> = (props) => {

    const activeRadio = props.active ? styles.activeHeader : styles.inactiveHeader;
    const checkedRadio = !props.active ? '' : styles.checkedRadio;

    return (
        <div className={`accordion-header ${styles.container}`}>
            <div className={`${styles.radio} ${activeRadio} ${checkedRadio}`} onClick={props.onClick}>
                <label>
                    <input
                        type="radio"
                        name={props.name}
                        defaultChecked={props.active}
                    />
                    <span>{props.title}</span>
                </label>
            </div>
            <div>
                {props.children}
            </div>
        </div>
    );
};
