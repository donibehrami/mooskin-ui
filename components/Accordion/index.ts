import Accordion, {Content} from './Accordion';

Accordion.Content = Content;

export default Accordion;
export {Content};
