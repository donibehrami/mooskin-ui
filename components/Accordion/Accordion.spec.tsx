import * as React from 'react';
import Accordion, {Content, Header} from './Accordion';

import { mount, render, shallow } from 'enzyme';

describe('Accordion', () => {

    test('renders Accordion properly according to snapshot', () => {

        Date.now = jest.fn(() => 1482363367071);

        const component = shallow(
            <Accordion id="5" className="mySwitch" style={{color: 'black'}}>
                <Content title="title1">Content1</Content>
                <Content title="title2" active>Content2</Content>
                <Content title="title3" className="classname">Content3</Content>
            </Accordion>
        );

        expect(component.find('Header').length).toBe(3);
        expect(component.find('Content').length).toBe(3);

        expect(component.find('Content').first().prop('children')).toBe('Content1');

        expect(component).toMatchSnapshot();
    });

    test('renders Content properly according to snapshot', () => {

        const component = shallow(
            <Content
                title="title1"
                style={{color: 'blue'}}
                active
            >
                Some Content
            </Content>
        );

        expect(component).toMatchSnapshot();
    });

    test('renders Header properly according to snapshot', () => {

        Date.now = jest.fn(() => 1482363367071);

        const component = shallow(
            <Header
                title="title1"
                onClick={() => null}
                active
            />
        );

        expect(component).toMatchSnapshot();
    });

    test('changes the state and displays appropriate content when one of the headers is clicked and back', () => {

        const component = mount(
            <Accordion>
                <Content title="title1">Content1</Content>
                <Content title="title2" active>Content2</Content>
            </Accordion>
        );

        expect(component.state('active')).toBe(1);
        expect(component.find('.accordion-header').first().find('input').prop('defaultChecked')).toBeFalsy();
        expect(component.find('.accordion-header').last().find('input').prop('defaultChecked')).toBeTruthy();
        expect(component.find('.accordion-content').first().hasClass('invisible')).toBeTruthy();
        expect(component.find('.accordion-content').last().hasClass('visible')).toBeTruthy();

        component.find('.radio').first().simulate('click');

        expect(component.state('active')).toBe(0);
        expect(component.find('.accordion-header').first().find('input').prop('defaultChecked')).toBeTruthy();
        expect(component.find('.accordion-header').last().find('input').prop('defaultChecked')).toBeFalsy();
        expect(component.find('.accordion-content').first().hasClass('visible')).toBeTruthy();
        expect(component.find('.accordion-content').last().hasClass('invisible')).toBeTruthy();

        component.find('.radio').last().simulate('click');

        expect(component.state('active')).toBe(1);
        expect(component.find('.accordion-header').first().find('input').prop('defaultChecked')).toBeFalsy();
        expect(component.find('.accordion-header').last().find('input').prop('defaultChecked')).toBeTruthy();
        expect(component.find('.accordion-content').first().hasClass('invisible')).toBeTruthy();
        expect(component.find('.accordion-content').last().hasClass('visible')).toBeTruthy();

    });
});
