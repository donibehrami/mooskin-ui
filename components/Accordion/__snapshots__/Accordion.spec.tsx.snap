// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders Accordion properly according to snapshot 1`] = `
<div
  className="accordion-component container"
>
  <Header
    active={false}
    name="1482363367071"
    onClick={[Function]}
    title="title1"
  >
    <Content
      active={false}
      title="title1"
    >
      Content1
    </Content>
  </Header>
  <Header
    active={true}
    name="1482363367071"
    onClick={[Function]}
    title="title2"
  >
    <Content
      active={true}
      title="title2"
    >
      Content2
    </Content>
  </Header>
  <Header
    active={false}
    name="1482363367071"
    onClick={[Function]}
    title="title3"
  >
    <Content
      active={false}
      className="classname"
      title="title3"
    >
      Content3
    </Content>
  </Header>
</div>
`;

exports[`Accordion renders Content properly according to snapshot 1`] = `
<div
  className="accordion-content content visible"
  style={
    Object {
      "color": "blue",
    }
  }
>
  Some Content
</div>
`;

exports[`Accordion renders Header properly according to snapshot 1`] = `
<div
  className="accordion-header container"
>
  <div
    className="radio activeHeader checkedRadio"
    onClick={[Function]}
  >
    <label>
      <input
        defaultChecked={true}
        type="radio"
      />
      <span>
        title1
      </span>
    </label>
  </div>
  <div />
</div>
`;
